src/main_menu.cpp

```diff
@@ -851,145 +856,154 @@ void main_menu::world_tab()
-            }
-            wrefresh( w_open );
-            refresh();
-            std::string action = ctxt.handle_input();
-
             if( action == "DOWN" ) {
                 if( sel3 > 0 ) {
                     --sel3;
                 } else {
-                    sel3 = all_worldnames.size() - 1;
+                    sel3 = vWorldSubItems.size() - 1;
                 }
+                on_move();
             } else if( action == "UP" ) {
-                if( sel3 < ( int )all_worldnames.size() - 1 ) {
+                if( sel3 < static_cast<int>( vWorldSubItems.size() ) - 1 ) {
                     ++sel3;
                 } else {
                     sel3 = 0;
                 }
+                on_move();
             } else if( action == "LEFT" || action == "QUIT" ) {
                 layer = 2;
-
-                print_menu( w_open, sel1, iMenuOffsetX, iMenuOffsetY );
             }
+
             if( action == "RIGHT" || action == "CONFIRM" ) {
-                if( sel3 >= 0 && sel3 < ( int )all_worldnames.size() ) {
+                if( sel3 == 2 ) { // Active World Mods
+                    WORLDPTR world = world_generator->get_world( all_worldnames[sel2 - 1] );
+                    world_generator->show_active_world_mods( world->active_mod_order );
+
+                } else {
                     bool query_yes = false;
                     bool do_delete = false;
-                    if( sel2 == 1 ) { // Delete World
+                    if( sel3 == 0 ) { // Delete World
                         if( query_yn( _( "Delete the world and all saves?" ) ) ) {
                             query_yes = true;
                             do_delete = true;
                         }
-                    } else if( sel2 == 2 ) { // Reset World
+                    } else if( sel3 == 1 ) { // Reset World
                         if( query_yn( _( "Remove all saves and regenerate world?" ) ) ) {
                             query_yes = true;
                             do_delete = false;
                         }
                     }
 
-                    layer = 2; // Go to world submenu, not list of worlds
                     if( query_yes ) {
-                        world_generator->delete_world( all_worldnames[sel3], do_delete );
+                        layer = 2; // Go to world submenu, not list of worlds
+
+                        world_generator->delete_world( all_worldnames[sel2 - 1], do_delete );
 
                         savegames.clear();
                         MAPBUFFER.reset();
                         overmap_buffer.clear();
 
-                        if( all_worldnames.empty() ) {
+                        if( do_delete ) {
                             sel2 = 0; // reset to create world selection
                         }
                     }
                 }
-                print_menu( w_open, sel1, iMenuOffsetX, iMenuOffsetY );
+            }
+        } else if( layer == 2 ) { // Show world names
```

Comment:
Those layer magic constants would be much more readable if you defined an enum for them.

Reply:
How would an enum calling them layer_1, layer_2, layer_3 make them any better?

Reply:
It wouldn't. You'd have to make a sane enum, with names like `world_names`, not `layer_2`.